### ISSUE TITLE
RFC: array-agnostic `quantile`

### ISSUE THREAD

**Issue by mdhaber** [data-apis/array-api#795]
I'm working on adding array API support in `scipy.stats` (scipy/scipy#20544) and one of the the things I'll need is a `quantile` function. If there is some support for this idea, I'll convert this issue into a proper proposal.

Looks like there is already wide support:
- [`numpy.quantile`](https://numpy.org/doc/stable/reference/generated/numpy.quantile.html)
- [`torch.quantile`](https://pytorch.org/docs/stable/generated/torch.quantile.html)
- [`cupy.quantile`](https://docs.cupy.dev/en/stable/reference/generated/cupy.quantile.html)
- [`jax.numpy.quantile`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.quantile.html)
- [`dask.dataframe.DataFrame.quantile`](https://docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.quantile.html)
- [`tfp.stats.quantiles`](https://www.tensorflow.org/probability/api_docs/python/tfp/stats/quantiles)
- [`xarray.DataArray.quantile`](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.quantile.html)

Previous discussions (not much):
- https://github.com/data-apis/array-api/issues/10#issuecomment-656228898
- https://github.com/data-apis/array-api/issues/232#issuecomment-884454859

There are many conventions for calculating quantiles. Only a few methods would be required by the standard, and if choice of a default is too contentious, perhaps the array-API can consider `method` to be a required keyword argument, and libraries would be welcome to keep their own default.

**Comment by rgommers** [comment]
Thanks for the proposal @mdhaber. I don't quite have an opinion yet - I think it in part depends on the situation with methods (see below).

Also - how much do you actually need this? I only count two instances of it being used in SciPy, one of which is a test case. I just did a grep, so I may be missing some dynamic usage perhaps. The one instance is pretty simple, no keyword usage:

https://github.com/scipy/scipy/blob/d55cb95282b5cfab381e373d3c127630c7e915a0/scipy/stats/tests/test_fit.py#L1002

> Only a few methods would be required by the standard, 

Which ones? Could we get away with only a default method, and hence no `method` keyword?

> and if choice of a default is too contentious, perhaps the array-API can consider `method` to be a required keyword argument, and libraries would be welcome to keep their own default.

**Comment by kgryte** [comment]
Related is our previous [discussion](https://github.com/data-apis/array-api/issues/10#issuecomment-656067265) on `median`, which we left out due to implementation difficulty in distributed contexts. That stated, [Dask](https://docs.dask.org/en/stable/generated/dask.array.median.html) does implement `median`, so may be time to revisit this.

**Comment by mdhaber** [comment]
> Also - how much do you actually need this?

Personally, I don't need it very badly. You're right that `quantile`/`percentile` only appear in a few `stats` functions (`stats.bootstrap`, `stats.iqr`, and some distribution fitting code), and it is easy to compute `quantiles` using existing array-API functions namely (`sort`). However, when only a few `percentiles` are needed, it is more efficient to use a partition approach. One could propose adding `partition`, since it is more general, but I would only use it for `quantile`.

> Which ones? Could we get away with only a default method, and hence no method keyword?
The standard reference is [Sample Quantiles in Statistical Packages](https://www.tandfonline.com/doi/abs/10.1080/00031305.1996.10473566). 

 > The current variation in sample quantile definitions causes confusion, and so there is a need to standardize the definition of sample quantile across packages and within packages... and we propose that $Q_8$ is the best choice

$Q_8$ corresponds with NumPy's `median_unbiased`. However, $Q_7$ (`linear`) is the default in most libraries. It would be unfortunate to have to choose between the two. $Q_5$ (`hazen`) also satisfies many desirable properties. So I suppose I would suggest having these three in that order of priority.

One wrench I'd like to throw into my own proposal: 
The current behavior of  [`np.quantile(a, q, axis)`](https://numpy.org/doc/stable/reference/generated/numpy.quantile.html) (and other implementations that follow NumPy) is for the quantiles specified by `q`, a 1D array-like, to be computed for *all* `axis`-slices of the data `a`. This makes it impossible to efficiently compute a different quantile for *each* axis-slice of `a` (which is needed, for example, in a vectorized implementation of the BCA-bootstrap). 
I would propose following normal broadcasting rules between `a` and `q` (possibly with the constraint that `q` must be 0D or have length `1` along `axis`), which would meet both needs. A hybrid possibility, which would just be an extension of the existing behavior, would be to allow `q` to be an N-D array that is broadcastable with `a` *except along `axis`*. In this case, for *each* pair of corresponding axis-slices `a_i` and `q_i`, it would compute *all* quantiles of `a_i` specified by `q_i`.

**Comment by kgryte** [comment]
I did a bit of digging, with the disclaimer that my digging is incomplete, but I'll try to summarize my initial findings below.

## Overview

I took a peak at which APIs were implemented across array libraries, spot-checked whether/how they were implemented, and did a search for which APIs were used in SciPy and sklearn.

## median

### Usage

- SciPy: [low usage](https://github.com/search?q=repo%3Ascipy%2Fscipy+np.median&type=code); more in tests, than in implementations
- scikit-learn: a reasonable amount of [usage](https://github.com/search?q=repo%3Ascikit-learn%2Fscikit-learn%20np.median&type=code)

### Implementations

- NumPy implements in terms of [`partition`](https://github.com/numpy/numpy/blob/d35cd07ea997f033b2d89d349734c61f5de54b0d/numpy/lib/function_base.py#L3931)
- CuPy implements as an element-wise kernel
- PyTorch returns the lower of two medians for an even number of elements, rather than the average (NumPy returns the avg to avoid breaking astropy); to return an average, must call `quantile`.
- Dask supports `median`
- TensorFlow does not appear to have a dedicated `median` API, instead deferring to `percentile`

## quantile

### Usage

- SciPy: [low usage](https://github.com/search?q=repo%3Ascipy%2Fscipy+np.quantile&type=code)
- scikit-learn: limited [usage](https://github.com/search?q=repo%3Ascikit-learn%2Fscikit-learn+np.quantile&type=code); mainly in tests

### Implementations

- NumPy implements in terms of [`partition`](https://github.com/numpy/numpy/blob/v1.26.0/numpy/lib/function_base.py#L4764)
- CuPy relies on [`sort`](https://github.com/cupy/cupy/blob/v13.1.0/cupy/_statistics/order.py#L219)
  - only supports a subset of methods: `lower`, `higher`, `midpoint`, `nearest`, and `linear`
- PyTorch supports the same methods as CuPy
- Dask supports in DataFrames, but not arrays
- TensorFlow implements in terms of [`percentile`](https://github.com/tensorflow/probability/blob/v0.23.0/tensorflow_probability/python/stats/quantiles.py#L645-L741)
  - supports the same methods as CuPy
  - API differs in that you specify the number of quantiles you want

## percentile

### Usage

- SciPy: [low usage](https://github.com/search?q=repo%3Ascipy%2Fscipy+np.percentile&type=code)
- scikit-learn: a reasonable amount of [usage](https://github.com/search?q=repo%3Ascikit-learn%2Fscikit-learn+np.percentile&type=code)

### Implementations

- NumPy implements in terms of [`quantile`](https://github.com/numpy/numpy/blob/d35cd07ea997f033b2d89d349734c61f5de54b0d/numpy/lib/function_base.py#L4283)
- PyTorch does not support
- CuPy implements in terms of [`quantile`](https://github.com/cupy/cupy/blob/v13.1.0/cupy/_statistics/order.py#L293)
  - supports only `linear`, `lower`, `higher`, `nearest`, and `midpoint` methods
- Dask supports "approximate" percentiles
  - supports same methods as CuPy
- TensorFlow implements in terms of [`sort`](https://github.com/tensorflow/probability/blob/v0.23.0/tensorflow_probability/python/stats/quantiles.py#L567)
  - supports same methods as CuPy

## partition

### Implementations

- NumPy: https://numpy.org/doc/stable/reference/generated/numpy.partition.html
- CuPy supports: https://docs.cupy.dev/en/stable/reference/generated/cupy.partition.html
- PyTorch does not support
- Dask does not support
- TensorFlow does not support

**Comment by mdhaber** [comment]
I reviewed `scipy.stats` more thoroughly for uses of `median`, `quantile`, and/or `percentile`. The first three functions would benefit from having `quantile` in the array API; for the rest, `median` would suffice.

- [`scipy.stats.bootstrap`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html) uses [`percentile` here](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_resampling.py#L643), but I would be happy to change it to use `quantile`. It would benefit from a `quantile` function with improved vectorization as described [above](https://github.com/data-apis/array-api/issues/795#issuecomment-2075456196).
- [`scipy.stats.iqr`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.iqr.html) uses [`percentile` here](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_stats_py.py#L3384), but I'd be happy to change it to use `quantile`.
- [`scipy.stats.epps_singleton_2samp`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.epps_singleton_2samp.html) uses [`iqr` here](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_hypotests.py#L115).
- [`scipy.stats.siegelslopes`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.siegelslopes.html) is currently implemented using Pythran, but I think it can be reasonably fast without Pythran if we wanted to use other backends. (Could Pythran dispatch to other backends in the future?) It uses [`median` here](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_stats_pythran.py#L167) and again outside the loop.
- [`scipy.stats.theilslopes`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.theilslopes.html) uses [`median` here.](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_stats_mstats_common.py#L340)
- [`scipy.stats.median_test`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.median_test.html), as one might imagine, uses [`median` here](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_morestats.py#L4288).
- [`scipy.stats.levene`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.levene.html) uses [`median` here](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_morestats.py#L3304).
- [`scipy.stats.fligner`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fligner.html) uses [`median` here](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_morestats.py#L3604).
- [`scipy.stats.mstats.sen_seasonal_slopes`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mstats.sen_seasonal_slopes.html) uses [`np.ma.median` here](https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_mstats_basic.py#L1333), but one could argue that this should be implemented in `stats` with regular `np.median`.

Most of these functions have something else that would make array API conversion challenging at the moment, but I don't think any have non-starters for array API support. I've omitted uses I don't think will get array API support any time soon (e.g. `levy_stable`). There are also uses in other sub-packages (e.g. [median_filter](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.median_filter.html)), but I'm less familiar with their prospects for array API support.

**Comment by lorentzenchr** [comment]
In my experience, quantiles are very important in practice and occur all over the places.

> > Which ones? Could we get away with only a default method, and hence no method keyword?
> > The standard reference is [Sample Quantiles in Statistical Packages](https://www.tandfonline.com/doi/abs/10.1080/00031305.1996.10473566).
> 
> > The current variation in sample quantile definitions causes confusion, and so there is a need to standardize the definition of sample quantile across packages and within packages... and we propose that Q8 is the best choice
> 
> Q8 corresponds with NumPy's `median_unbiased`. However, Q7 (`linear`) is the default in most libraries. It would be unfortunate to have to choose between the two. Q5 (`hazen`) also satisfies many desirable properties. So I suppose I would suggest having these three in that order of priority.

I could imagine to let the default method vary (for easier consensus), but each one must have a `linear` option (most have it already).
For weighted quantiles in particular, method `inverted_cdf` would be mandatory.

**Comment by ogrisel** [comment]
>  For weighted quantiles in particular, method inverted_cdf would be mandatory.

I would also advocate for `averaged_inverted_cdf` `in addition to `inverted_cdf`, which makes it possible to get a strict equivalence between unweighted quantiles (typically computed with the `linear` strategy by default) and unit-weighted quantiles on even sample sizes.

**Comment by mdhaber** [comment]
@rgommers you were concerned about meetings. I wanted to check whether it was clear that these methods differ only in the choice of two numbers, as described in NumPy's [`quantile`](https://numpy.org/doc/2.1/reference/generated/numpy.quantile.html#re01cd3f3acfe-1) documentation, so they are not really distinct implementations (as e.g. Harwell-Davis quantiles would be).

If it helps, the methods I suggested initially can be parametrized by two floats in $[0, 1]$, so one option would be to expose those floats as as parameters (like [`mstats.mquantiles`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mstats.mquantiles.html)) does. For the rest, the two numbers are functions of the quantile `q` and number of observations `n`, so it's easier to attach names to them.

OTOH, at this point I'm leaning toward adding an implementation to SciPy in terms of array API functions, the key operations being `sort` and `take_along_axis` IIRC. It would probably be slower than a backend-specific implementation in some cases because it would always do a full sort, but I suppose we always have the option of delegating.

**Comment by lorentzenchr** [comment]
> adding an implementation to SciPy in terms of array API functions

@mdhaber You might want to contact/coordinate with scikit-learn for that. (Apart from numpy 2.0 not yet being the minimal dependency) The array API compatibility is meanwhile one of the main motivations to have our own implementation of it, see the recent PR https://github.com/scikit-learn/scikit-learn/pull/29034.

All technical details aside, what is the right approach to move this RFC forward? How/Where is the decision making body, etc.?

**Comment by mdhaber** [comment]
That's fine if scikit-learn needs its own private function. We'll need something (public or private) in SciPy anyway. I opened gh-22352 about that.

**Comment by jakirkham** [comment]
Linking `topk` (or `top_k`) discussion as this is a somewhat related question: https://github.com/data-apis/array-api/issues/629

Note with Dask users are often recommended to use something like `topk` as `median`, `quantile`, etc. requiring sorting large data, which can be complicated

**Comment by mdhaber** [comment]
gh-22352 added [`scipy.stats.quantile`](https://scipy.github.io/devdocs/reference/generated/scipy.stats.quantile.html#scipy.stats.quantile), which is tested with NumPy, CuPy, and PyTorch right now, and we can probably add support for the other backends in time. I think this will meet the ecosystem's needs for median and quantiles. Performance might be improved with `argpartition` (gh-629), so I'll close this in favor of that.

**Comment by lorentzenchr** [comment]
@lucascolley Out of interest: Who decided when that this is *not planned*?

**Comment by lucascolley** [comment]
![Image](https://github.com/user-attachments/assets/5d28df68-f5e1-44e4-b7b9-0afeef94dd5f)
we can reopen if you think discussion should continue! I closed it more so as 'stale' than 'not planned', both of which are less misleading than 'completed' IMO

**Comment by mdhaber** [comment]
I opened the issue, and my concern is _resolved_ (even if the resolution is in another library), so I'll close again. Anyone is welcome to open a new issue and link to this one.

**Comment by lucascolley** [comment]
Okay, I edited the title to avoid misleading that `quantile` was added to the standard

**Comment by leofang** [comment]
If it's not added, and there is no plan to add it, what's wrong with "not planned"?

**Comment by mdhaber** [comment]
I don't have a problem with that, either. It is also accurate to say that it is "not planned".

Perhaps there is confusion about the difference between "not planned" and "decided against adding"?

**Comment by lucascolley** [comment]
> Perhaps there is confusion about the difference between "not planned" and "decided against adding"?

It looks like we can use the label ~"status: Rejected" to make this distinction
